nitro-protocol/contracts/CurrencySwapApp.sol

```diff
@@ -0,0 +1,175 @@
+        //         return;
+        //     }
+        //     revert('bad candidate turnNum; |proof|=0');
+        // }
+        // // State 2 can be supported via a forced transition from state 1:
+        // //
+        // //      (2)_B     [redemption state signed by Bob, includes a voucher signed by Alice. The outcome may be updated in favour of Bob]
+        // //      ^
+        // //      |
+        // //      (1)_AIB   [fully signed postfund]
+        // if (proof.length == 1) {
+        //     requireProofOfUnanimousConsensusOnPostFund(proof[0], fixedPart.participants.length);
+        //     require(candidate.variablePart.turnNum == 2, 'bad candidate turnNum; |proof|=1');
+        //     require(
+        //         NitroUtils.isClaimedSignedBy(candidate.signedBy, 2),
+        //         'redemption not signed by Bob'
+        //     );
+        //     uint256 voucherAmount = requireValidVoucher(candidate.variablePart.appData, fixedPart);
+        //     requireCorrectAdjustments(
+        //         proof[0].variablePart.outcome,
+        //         candidate.variablePart.outcome,
+        //         voucherAmount
+        //     );
+        //     return;
+        // }
+        // revert('bad proof length');
+    }
+
+    function requireProofOfUnanimousConsensusOnPostFund(
+        RecoveredVariablePart memory rVP,
+        uint256 numParticipants
+    ) internal pure {
+        require(rVP.variablePart.turnNum == 1, 'bad proof[0].turnNum; |proof|=1');
+        require(
+            NitroUtils.getClaimedSignersNum(rVP.signedBy) == numParticipants,
+            'postfund !unanimous; |proof|=1'
+        );
+    }
+
+    function requireValidVoucher(
+        bytes memory appData,
+        FixedPart memory fixedPart
+    ) internal pure returns (uint256) {
+        VoucherAmountAndSignature memory voucher = abi.decode(appData, (VoucherAmountAndSignature));
+
+        address signer = NitroUtils.recoverSigner(
+            keccak256(abi.encode(NitroUtils.getChannelId(fixedPart), voucher.amount)),
+            voucher.signature
+        );
+        //TODO: Verify amount, signature of Maker
+        require(signer == fixedPart.participants[0], 'invalid signature for voucher'); // could be incorrect channelId or incorrect signature
+        return voucher.amount;
+    }
+
+    function requireCorrectAdjustments(
+        Outcome.SingleAssetExit[] memory oldOutcome,
+        Outcome.SingleAssetExit[] memory newOutcome,
+        uint256 voucherAmount
+    ) internal pure {
+        // TODO: require Outcome.length == 2;
+        // && Outcome[0].asset == TakerAsset; Outcome[1].asset == MakerAsset
+        require(
+            oldOutcome.length == 1 &&
+                newOutcome.length == 1 &&
+                oldOutcome[0].asset == address(0) &&
+                newOutcome[0].asset == address(0),
+            'only native asset allowed'
+        );
+
```

Comment:
@nksazonov We can apply the respective amount of margin to the allocation amount
for simplicity, I suggest we express margin in amount - margin
the purpose is to cumulate margin calls,

amount - 10, -20, -30, -10.